src/Data/Formatter/Parser/Interval.purs

```diff
@@ -0,0 +1,80 @@
+module Data.Formatter.Parser.Interval
+  ( parseRecurringInterval
+  , parseInterval
+  , parseIsoDuration
+  , parseDateTime
+  , extendedDateTimeFormatInUTC
+  ) where
+
+import Prelude
+
+import Control.Alt ((<|>))
+import Data.DateTime (DateTime)
+import Data.Either (Either(..), fromRight)
+import Data.Foldable (class Foldable, fold, foldMap, intercalate)
+import Data.Formatter.DateTime (unformatParser, Formatter, parseFormatString)
+import Data.Formatter.Parser.Number (parseNumber, parseMaybeInteger)
+import Data.Interval as I
+import Data.Interval.Duration.Iso (IsoDuration, mkIsoDuration, prettyError)
+import Data.Maybe (Maybe)
+import Data.Monoid (class Monoid, mempty)
+import Data.Traversable (sequence)
+import Data.Tuple (Tuple(..), snd)
+import Partial.Unsafe (unsafePartial)
+import Text.Parsing.Parser as P
+import Text.Parsing.Parser.Combinators as PC
+import Text.Parsing.Parser.String as PS
+
+parseRecurringInterval ∷ ∀ a b. P.Parser String a → P.Parser String b → P.Parser String (I.RecurringInterval a b)
+parseRecurringInterval duration date =
+  I.RecurringInterval <$> (PS.string "R" *> parseMaybeInteger) <*> (PS.string "/" *> parseInterval duration date)
+
+parseInterval ∷ ∀ a b. P.Parser String a → P.Parser String b → P.Parser String (I.Interval a b)
+parseInterval duration date = [startEnd, durationEnd, startDuration, durationOnly] <#> PC.try # PC.choice
+  where
+  startEnd = I.StartEnd <$> date <* PS.string "/" <*> date
+  durationEnd = I.DurationEnd <$> duration <* PS.string "/" <*> date
+  startDuration = I.StartDuration <$> date <* PS.string "/" <*> duration
+  durationOnly = I.DurationOnly <$> duration
+
+parseIsoDuration ∷ P.Parser String IsoDuration
+parseIsoDuration = do
+  dur ← parseDuration
+  case mkIsoDuration dur of
+    Left errs → let errorStr = intercalate ", " (prettyError <$> errs)
+      in P.fail $ "Extracted Duration is not valid ISO duration (" <> errorStr <> ")"
+    Right a → pure a
+
+parseDuration ∷ P.Parser String I.Duration
+parseDuration = PS.string "P" *> (weekDuration <|> fullDuration)
+  where
+  weekDuration = mkComponentsParser [ Tuple I.week "W" ]
+  fullDuration = (append <$> durationDatePart <*> durationTimePart) `notEmpty` "Must contain valid duration components"
+  durationDatePart = PC.option mempty $ PC.try $ mkComponentsParser [ Tuple I.year "Y" , Tuple I.month "M" , Tuple I.day "D" ]
+  durationTimePart = PC.option mempty $ (PC.try $ PS.string "T") *> (mkComponentsParser [ Tuple I.hour "H" , Tuple I.minute "M" , Tuple I.second "S" ])
+
+
+notEmpty ∷ ∀ a. Monoid a ⇒ Eq a ⇒ P.Parser String a → String → P.Parser String a
+notEmpty p str = p >>= \x → if x == mempty then P.fail str else pure x
+
+mkComponentsParser ∷ Array (Tuple (Number → I.Duration) String) → P.Parser String I.Duration
+mkComponentsParser arr = p `notEmpty` ("None of valid duration components (" <> (show $ snd <$> arr) <> ") were present")
```

Comment:
I'd suggest using operator instead of infix function here. Like `<?>` in strongcheck. Or prefix func application. It's a bit mistyfying in that case, I mean infix `notEmpty`, maybe `msgIfEmpty` or something else.
Compare:
```purescript
p <:?:> "Critical Error!!!" 
nonEmptyParser p "Critical Error!!!"
p `messageIfInputIsEmpty` "Critical Error!!!"
```